### PR TITLE
APEXCORE-408: Ability to schedule Sub-DAG from running application.

### PR DESCRIPTION
**Pull request by tushargosavi**
Pull request for dynamic dag modification through stats listener.  It provides following
functionality
- StatsListener can access the opearator name for easily detecting which opearator stats are being processed.
- StatsListener can create a instance of object through which it can submit dag modifications to the engine.
- StatsListener can return dag changes as a response to engine.
- PlanModifier is modified to take a DAG and apply it on the existing running DAG and deploy the changes.

The following functionality is not working yet.
- The new opearator does not start from the correct windowId (https://issues.apache.org/jira/browse/APEXCORE-532)
- Relanched application failed to start when it was killed after dynamic dag modification.
- There is no support for resuming operator from previous state when they were removed. This could be achived through
  readig state through external storage on setup.
- persist operator support is not present for newly added streams.
- Not all parts are covered through tests.

The demo application using the feature is available at
https://github.com/tushargosavi/apex-dynamic-scheduling

There are two variations of WordCount application. The first variation detects the presence of
new files start a disconnected DAG to process the data.
(https://github.com/tushargosavi/apex-dynamic-scheduling/blob/master/src/main/java/com/datatorrent/wordcount/WordCountApp.java)

The second application (https://github.com/tushargosavi/apex-dynamic-scheduling/blob/master/src/main/java/com/datatorrent/wordcount/ExtendApp.java),
initially only one reader operator is running in the DAG, and provides pendingFiles as auto-metric to stat listener.
On detecting pending files it attaches splitter counter and output operator to the read operator. Once files are processed the splitter, counter and
output operators are removed and added back again if new data files are added into the directory.

### CODE REVIEW

File: api/src/main/java/com/datatorrent/api/DAG.java [hunk 1]
```diff
@@ -288,4 +288,23 @@
...
+
+  /**
+   * This class provides functionality to extend existing DAG. The type of extensions allowed are
+   *
+   * <li>Add a new operator</li>
+   * <li>Remove an existing operator</li>
+   * <li>Add an operator to stream</li>
+   * <li>Add an stream between existing operator and new opeartor</li>
+   */
+  interface DAGChangeSet extends DAG
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/DAG.java]: Why does it extend DAG?

File: api/src/main/java/com/datatorrent/api/StatsListener.java [hunk 2]
```diff
@@ -115,6 +117,55 @@
     List<OperatorResponse> getOperatorResponse();
   }
 
+  /**
+   * An interface to the DAG. Stats listener can get information about
+   * operator or other elements in the DAG through this interface. currerntly
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/StatsListener.java]: check all doc blocks for spelling

File: api/src/main/java/com/datatorrent/api/StatsListener.java [hunk 3]
```diff
@@ -115,6 +117,55 @@
...
+   * we only provide method to extract the operator name based on the physical
+   * id of the operator. In future more methods can be added.
+   *
+   */
+  interface StatsListenerContext
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/StatsListener.java]: How does this converge with existing processStats?

File: api/src/main/java/com/datatorrent/api/StatsListener.java [hunk 4]
```diff
@@ -115,6 +117,55 @@
...
+  }
+
+  /**
+   * If StatsListener implements ContextAwareStatsListener interface, then engine will
+   * provide a reference to StatsListenerContext using which listener can examine
+   * current state of the DAG.
+   */
+  interface ContextAwareStatsListener
+  {
+    void setContext(StatsListenerContext context);
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/StatsListener.java]: Any reason to not provide this as argument to processStats?

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java [hunk 5]
```diff
@@ -2904,6 +2905,8 @@ public LogicalPlan getLogicalPlan()
     return future;
   }
 
+
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: remove extra modifications

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java [hunk 6]
```diff
@@ -2918,13 +2921,7 @@ public Object call() throws Exception
...
       // clone logical plan, for dry run and validation
       LOG.info("Begin plan changes: {}", requests);
-      LogicalPlan lp = plan.getLogicalPlan();
-      ByteArrayOutputStream bos = new ByteArrayOutputStream();
-      LogicalPlan.write(lp, bos);
-      bos.flush();
-      ByteArrayInputStream bis = new ByteArrayInputStream(bos.toByteArray());
-      lp = LogicalPlan.read(bis);
-
+      LogicalPlan lp = plan.cloneLogicalPlan();
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: SerializationUtils.clone(...) ?

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java [hunk 7]
```diff
@@ -3017,6 +3014,7 @@ public static StreamingContainerManager getInstance(RecoveryHandler rh, LogicalP
 
         // restore checkpoint info
         plan.syncCheckpoints(scm.vars.windowStartMillis, scm.clock.getTime());
+        plan.setContextAfterDAGRecovery();
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: This would not be needed when passing context with processStats.

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java [hunk 8]
```diff
@@ -3208,10 +3206,50 @@ public Object call() throws Exception
...
+        pm.applyDagChangeSet(request);
+        lp.validate();
+
+        LOG.info("validated logical plan, now making actual change");
+        pm = new PlanModifier(plan);
+        pm.applyDagChangeSet(request);
+        LOG.info("plan change done");
+        return null;
+      } finally {
+        plan.clearPendingChangeRequest();
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: why is this needed?

File: engine/src/main/java/com/datatorrent/stram/plan/logical/mod/DAGChangeSetImpl.java [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+import com.datatorrent.api.DAG.DAGChangeSet;
+import com.datatorrent.api.Operator;
+import com.datatorrent.stram.plan.logical.LogicalPlan;
+
+/**
+ * An implementation for DAGChangeSet. Instance of this object will be provided
+ * to StatsListener through context, stat listener can use this object to modify
+ * existing DAG and return modified DAG to engine.
+ */
+public class DAGChangeSetImpl extends LogicalPlan implements DAGChangeSet
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/logical/mod/DAGChangeSetImpl.java]: Why is the change set not separate from DAG. Exsisting plan modification from the CLI works without such extension.

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java [hunk 10]
```diff
@@ -1898,4 +1967,87 @@ public Integer getStreamCodecIdentifier(StreamCodec<?> streamCodecInfo)
...
+    ByteArrayOutputStream bos = new ByteArrayOutputStream();
+    LogicalPlan.write(lp, bos);
+    bos.flush();
+    ByteArrayInputStream bis = new ByteArrayInputStream(bos.toByteArray());
+    lp = LogicalPlan.read(bis);
+    return lp;
+  }
+
+
+  public void clearPendingChangeRequest()
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: Not clear why it is needed. Generally public methods should have javadoc.

**Review by tweise — CHANGES_REQUESTED** [review]

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java (hunk 5 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: done

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java (hunk 6 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: done

File: api/src/main/java/com/datatorrent/api/StatsListener.java (hunk 4 above)
Comment (tushargosavi) [reply api/src/main/java/com/datatorrent/api/StatsListener.java]: Adding a new argument will break binary compatibility of existing stats listeners. hence went for this approach. This could be implemented with
- Providing context through BatchedOperatorStats.
- Adding a new interface extending StatsListener which defines processStats with additional argument, for this listeners new method will be called by platform. please suggest how can we implement this?

File: api/src/main/java/com/datatorrent/api/DAG.java (hunk 1 above)
Comment (tushargosavi) [reply api/src/main/java/com/datatorrent/api/DAG.java]: My idea was to use the same api to specify dag changes with
addition of new methods to change existing DAG. This also
allows running existing application as a subdag, for
example following application run three applications one after
another, individual applications can be run separately too

https://github.com/tushargosavi/apex-dynamic-scheduling/blob/master/src/main/java/com/datatorrent/wordcount/lindag/DagSchedulingApp.java#L35

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java (hunk 8 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: At a time only one pending dag change request will be active. once the dag change request is processed successfully or unsuccessfully, it will be cleared so that new request can be submitted from statsListener. Stats listener has access to the feature through with it can check the status of last submitted dag change request.

File: api/src/main/java/com/datatorrent/api/StatsListener.java (hunk 2 above)
Comment (tushargosavi) [reply api/src/main/java/com/datatorrent/api/StatsListener.java]: done

File: api/src/main/java/com/datatorrent/api/StatsListener.java (hunk 3 above)
Comment (tushargosavi) [reply api/src/main/java/com/datatorrent/api/StatsListener.java]: It will be additional interface between DAG and statlistener. It can be used to get additional information about the operator/dag when processStats is called. currently processStats BatchedOperatorStats argument is very limited, in case of shared stats listener it is difficult to know for which operator processStats is being called. In presence of context stats listener can get operator name from operator id to determine the operator. Also it can provide additional interaction between engine and statslistener such as requesting dag change or shutting down the application.

File: api/src/main/java/com/datatorrent/api/StatsListener.java (hunk 4 above)
Comment (tushargosavi) [reply api/src/main/java/com/datatorrent/api/StatsListener.java]: I have changed the processStats to accept additional argument by extending the interface to maintain backward compatibility. But this solution require us to implement two processStats functional one from old interface and one from new interface.

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java (hunk 7 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: with additional parameter it is not needed. But let me know which is the preferred way?

File: engine/src/main/java/com/datatorrent/stram/plan/logical/mod/DAGChangeSetImpl.java (hunk 9 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/plan/logical/mod/DAGChangeSetImpl.java]: Existing plan modification use different api to change the dag. As I am using same API with few additions (for example addSink to existing stream/remove operator) to specify DAG change, I have extended current DAG implementation (LogicalPlan) to support additional methods.

File: engine/src/test/java/com/datatorrent/stram/engine/InputOperatorTest.java [hunk 11]
```diff
@@ -51,10 +52,13 @@
 
   public static class EvenOddIntegerGeneratorInputOperator implements InputOperator, com.datatorrent.api.Operator.ActivationListener<OperatorContext>
   {
+    @OutputPortFieldAnnotation(optional = true)
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/engine/InputOperatorTest.java]: Is this still needed?

File: api/src/main/java/com/datatorrent/api/DAG.java [hunk 12]
```diff
@@ -288,4 +288,23 @@
...
+  {
+    void removeOperator(String name);
+
+    void removeStream(String name);
+
+    StreamMeta extendStream(String id, Operator.InputPort... ports);
+
+    StreamMeta addStream(String id, String operatorName, String portName, Operator.InputPort... ports);
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/DAG.java]: What's the reason to not specify the source as port object?

File: api/src/main/java/com/datatorrent/api/DAG.java [hunk 13]
```diff
@@ -288,4 +288,23 @@
   {
 
   }
+
+  /**
+   * This class provides functionality to extend existing DAG. The type of extensions allowed are
+   *
+   * <li>Add a new operator</li>
```
Comment (tweise) [inline api/src/main/java/com/datatorrent/api/DAG.java]: Remove "a" and "an"

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java [hunk 14]
```diff
@@ -3199,13 +3194,40 @@ public Object call() throws Exception
...
+
+    @Override
+    public Object call() throws Exception
+    {
+      if (request == null) {
+        return null;
+      }
+
+      try {
+        LOG.info("starting plan change ");
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: "logical plan change"

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java [hunk 15]
```diff
@@ -3199,13 +3194,40 @@ public Object call() throws Exception
...
+        lp.validate();
+
+        LOG.info("validated logical plan, now making actual change");
+        pm = new PlanModifier(plan);
+        pm.applyDagChangeSet(request);
+        LOG.info("plan change done");
+        return null;
+      } finally {
+        /* At a time only one request can be active, once request is processed with or without error,
+         * clear the pendingDagChagneRequest so that new request can be submitted from the StatsListenerContext. */
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: spelling

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java [hunk 16]
```diff
@@ -1523,19 +1544,33 @@ public LogicalPlan getLogicalPlan()
...
+    }
     return this.logicalToPTOperator.get(logicalOperator).getAllOperators();
   }
 
   public List<PTOperator> getLeafOperators()
   {
     List<PTOperator> operators = new ArrayList<>();
     for (OperatorMeta opMeta : dag.getLeafOperators()) {
-      operators.addAll(getAllOperators(opMeta));
+      Collection<PTOperator> allOpers = getAllOperators(opMeta);
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java:1612]: why this?

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java [hunk 17]
```diff
@@ -1675,10 +1710,36 @@ public final void addLogicalOperator(OperatorMeta om)
     } else {
       initPartitioning(pnodes, 0);
     }
+    updateUpstreamsOutputMappings(om, null);
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: Spelling. Also not clear why it is needed. Is it to fix an existing bug or is it for newly introduced functionality?

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java [hunk 18]
```diff
@@ -1794,7 +1855,12 @@ public void setAvailableResources(int memoryMB)
   public void onStatusUpdate(PTOperator oper)
   {
     for (StatsListener l : oper.statsListeners) {
-      final StatsListener.Response rsp = l.processStats(oper.stats);
+      final StatsListener.Response rsp;
+      if (l instanceof StatsListener.StatsListenerWithContext) {
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: Would it be possible to internally handle all listeners through the new interface and adapt the old interface through adapter? We should also deprecate the old interface.

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PlanModifier.java [hunk 19]
```diff
@@ -280,4 +287,121 @@ public void applyChanges(PhysicalPlan.PlanContext physicalPlanContext)
...
+      LOG.info("Removing operator {}", name);
+      removeOperator(name);
+    }
+
+    if (physicalPlan != null) {
+      physicalPlan.deployChanges();
+    }
+  }
+
+  StreamMeta getSteramBySource(String operatorName, String portName)
```
Comment (tweise) [inline engine/src/main/java/com/datatorrent/stram/plan/physical/PlanModifier.java]: spelling

File: engine/src/test/java/com/datatorrent/stram/engine/InputOperatorTest.java [hunk 20]
```diff
@@ -51,10 +52,13 @@
...
     public final transient DefaultOutputPort<Integer> even = new DefaultOutputPort<Integer>();
+    @OutputPortFieldAnnotation(optional = true)
```
Comment (tweise) [inline engine/src/test/java/com/datatorrent/stram/engine/InputOperatorTest.java]: ditto

**Review by tweise — CHANGES_REQUESTED** [review]

File: engine/src/test/java/com/datatorrent/stram/engine/InputOperatorTest.java (hunk 11 above)
Comment (tushargosavi) [reply engine/src/test/java/com/datatorrent/stram/engine/InputOperatorTest.java]: removed

File: api/src/main/java/com/datatorrent/api/DAG.java (hunk 13 above)
Comment (tushargosavi) [reply api/src/main/java/com/datatorrent/api/DAG.java]: removed

File: api/src/main/java/com/datatorrent/api/DAG.java (hunk 12 above)
Comment (tushargosavi) [reply api/src/main/java/com/datatorrent/api/DAG.java]: While applying the changes the changes are applied on the cloned logical plan first avoid any invalid operations on current plan. Only after the validation is successful the current plan is changed. The port objects are different in the cloned plan hence could not use the reference to port object directly.

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java (hunk 14 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: done

File: engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java (hunk 15 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/StreamingContainerManager.java]: done

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java (hunk 16 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java:1612]: getAllOperators can return null when dag change in progress as logical plan is changed first and then physical plan is changed. This check is added to prevent null pointer exception during rest api calls when api tries to get physical operators for a logical operator but physical plan is not changed yet.

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java (hunk 17 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: This is required to handle case where an new operator's port is connected to existing operator output port with new stream. As during modifications only the new operator will get added but we need to change output stream mapping for existing operator.

Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: I will add the comment in the code

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java (hunk 18 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: working on it, will update pull request after it.

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PlanModifier.java (hunk 19 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/plan/physical/PlanModifier.java]: done

File: engine/src/test/java/com/datatorrent/stram/engine/InputOperatorTest.java (hunk 20 above)
Comment (tushargosavi) [reply engine/src/test/java/com/datatorrent/stram/engine/InputOperatorTest.java]: done

File: engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java (hunk 18 above)
Comment (tushargosavi) [reply engine/src/main/java/com/datatorrent/stram/plan/physical/PhysicalPlan.java]: Pushed one change to internally handled all listeners through new interface. operators implementing  the interface needs implement Serializable this might break backward compatibility as old operators implementing stats listeners will fail to serialize.  checking for another approach.

**Comment by tushargosavi** [comment]
@tweise I have updated the pull request

File: api/src/main/java/com/datatorrent/api/DAG.java (hunk 12 above)
Comment (tweise) [reply api/src/main/java/com/datatorrent/api/DAG.java]: As discussed on the call, it would be good to adopt a more object oriented and encapsulated approach. It will be good to handle the changes within the plan instead of passing around disjoined DAG fragments that require the use of String references. Is that doable?

**Comment by tweise** [comment]
@tushargosavi are you planning to make changes that will remove the user facing distinction between initial DAG and modification at runtime?

**Comment by tushargosavi** [comment]
@tweise, have done some prototype based on existing approach which does not require changes to DAG with limitation that new dag needs to be a disconnected Sub-DAG.

**Comment by tushargosavi** [comment]
closing this, working on implementation with different strategy.
